server/resp/src/main/java/org/infinispan/server/resp/Resp3Handler.java

```diff
@@ -43,313 +24,45 @@ protected void setCache(AdvancedCache<byte[], byte[]> cache) {
       ignorePreviousValueCache = cache.withFlags(Flag.SKIP_CACHE_LOAD, Flag.IGNORE_RETURN_VALUES);
    }
 
-   protected static final BiConsumer<byte[], ByteBufPool> GET_BICONSUMER = (innerValueBytes, alloc) -> {
-      if (innerValueBytes != null) {
-         bytesToResult(innerValueBytes, alloc);
-      } else {
-         stringToByteBuf("$-1\r\n", alloc);
-      }
-   };
-
-   protected static final BiConsumer<Object, ByteBufPool> OK_BICONSUMER = (ignore, alloc) ->
-         alloc.acquire(OK.length).writeBytes(OK);
-
-   protected static final BiConsumer<SetResponse, ByteBufPool> SET_BICONSUMER = (res, alloc) -> {
-      // The set operation has three return options, with a precedence:
-      //
-      // 1. Previous value or `nil`: when `GET` flag present;
-      // 2. `OK`: when set operation succeeded
-      // 3. `nil`: when set operation failed, e.g., tried using XX or NX.
-      if (res.isReturnValue()) {
-         GET_BICONSUMER.accept(res.value(), alloc);
-         return;
-      }
-
-      if (res.isSuccess()) {
-         OK_BICONSUMER.accept(res, alloc);
-         return;
-      }
-
-      GET_BICONSUMER.accept(null, alloc);
-   };
-
-   protected static final BiConsumer<Long, ByteBufPool> LONG_BICONSUMER = Resp3Handler::handleLongResult;
-
-   protected static final BiConsumer<byte[], ByteBufPool> DELETE_BICONSUMER = (prev, alloc) ->
-         stringToByteBuf(":" + (prev == null ? "0" : "1") + "\r\n", alloc);
-
    @Override
    protected CompletionStage<RespRequestHandler> actualHandleRequest(ChannelHandlerContext ctx, RespCommand type, List<byte[]> arguments) {
-      switch (type) {
-         case PING:
-            if (arguments.size() == 0) {
-               stringToByteBuf("$4\r\nPONG\r\n", allocatorToUse);
-               break;
-            }
-            // falls-through
-         case ECHO:
-            byte[] argument = arguments.get(0);
-            ByteBuf bufferToWrite = stringToByteBufWithExtra("$" + argument.length + "\r\n", allocatorToUse, argument.length + 2);
-            bufferToWrite.writeBytes(argument);
-            bufferToWrite.writeByte('\r').writeByte('\n');
-            break;
-         case SET:
-            if (arguments.size() != 2) {
-               return stageToReturn(SetOperation.performOperation(cache, arguments), ctx, SET_BICONSUMER);
-            }
-            return stageToReturn(ignorePreviousValueCache.putAsync(arguments.get(0), arguments.get(1)), ctx, OK_BICONSUMER);
-         case GET:
-            byte[] keyBytes = arguments.get(0);
+      if (type instanceof BufferResponseCommand) {
```

Comment:
While I see this will evolve into a simple perform call. We need to pay attention to the fact that this will be a megamorphic call. We definitely need to do performance numbers, but it might be worth adding an enum type for optimized calls that use the previous switch and others just use the perform method.

Reply:
I did a quick check locally with redis-benchmark `-t set,get -d 9000 -r 1000 -n 500000`, comparing the current version and another which check if it is `GET` or `SET`. Three runs, restarting the server between checks:

Current version:

```text
  throughput summary: 74404.77 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.349     0.096     0.335     0.479     0.583    10.199

  throughput summary: 75086.35 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.346     0.096     0.335     0.447     0.543     6.831

  throughput summary: 75918.62 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.342     0.096     0.327     0.447     0.535    11.175
```

With `GET` and `SET` checks:

```
  throughput summary: 74526.75 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.349     0.104     0.335     0.463     0.551     9.103


  throughput summary: 75723.16 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.343     0.080     0.335     0.447     0.527     8.935

  throughput summary: 73035.34 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.356     0.072     0.343     0.479     0.575     6.255
```

I used `instanceof` for the cast, no enum checking.